crates/sui-bridge-indexer/src/sui_checkpoint_ingestion.rs

```diff
@@ -0,0 +1,302 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::config::Config;
+use crate::metrics::BridgeIndexerMetrics;
+use crate::postgres_manager::PgPool;
+use crate::schema::progress_store::{columns, dsl};
+use crate::sui_worker::SuiBridgeWorker;
+use crate::{models, schema};
+use async_trait::async_trait;
+use diesel::dsl::now;
+use diesel::{
+    delete, ExpressionMethods, OptionalExtension, QueryDsl, RunQueryDsl, SelectableHelper,
+};
+use mysten_metrics::spawn_monitored_task;
+use std::cmp::min;
+use sui_data_ingestion_core::{
+    DataIngestionMetrics, IndexerExecutor, ProgressStore, ReaderOptions, WorkerPool,
+};
+use sui_types::base_types::TransactionDigest;
+use sui_types::messages_checkpoint::CheckpointSequenceNumber;
+use tokio::sync::oneshot;
+use tokio::sync::oneshot::Sender;
+
+pub struct SuiCheckpointSyncer {
+    pool: PgPool,
+    bridge_genesis_checkpoint: u64,
+}
+
+impl SuiCheckpointSyncer {
+    pub fn new(pool: PgPool, bridge_genesis_checkpoint: u64) -> Self {
+        // read all task from db
+        SuiCheckpointSyncer {
+            pool: pool.clone(),
+            bridge_genesis_checkpoint,
+        }
+    }
+    pub async fn start(
+        self,
+        config: &Config,
+        indexer_meterics: BridgeIndexerMetrics,
+        ingestion_metrics: DataIngestionMetrics,
+    ) -> anyhow::Result<(), anyhow::Error> {
+        // Update tasks first
+        let tasks = self.tasks()?;
+        // checkpoint workers
+        match tasks.latest_checkpoint_task() {
+            None => {
+                // No task in database, start latest checkpoint task and backfill tasks
+                // if resume_from_checkpoint, use it for the latest task, if not set, use bridge_genesis_checkpoint
+                let start_from_cp = config
+                    .resume_from_checkpoint
+                    .unwrap_or(self.bridge_genesis_checkpoint);
+                self.register_task(new_task_name(), start_from_cp, i64::MAX)?;
+
+                // Create backfill tasks
+                if start_from_cp != config.bridge_genesis_checkpoint {
+                    let mut current_cp = self.bridge_genesis_checkpoint;
+                    while current_cp < start_from_cp {
+                        let target_cp = min(current_cp + config.back_fill_lot_size, start_from_cp);
```

Comment:
I keep wondering whether we should just define ranges/tasks in the config/db or whether we should generate multiple tasks as we do here.
We talked about it and it's possible to work with both (I think) but it seems the code would be simpler if we just started ranges as we are told